server/src/Store.hs

```diff
@@ -1,22 +1,60 @@
+{-# LANGUAGE OverloadedStrings #-}
 module Store
 (
+  Modification (..),
   Path,
+  Value,
+  modificationPath,
+  applyModification,
   delete,
   insert,
   lookup,
   lookupOrNull,
 )
 where
 
-import Data.Aeson (Value (..))
+import Data.Aeson (Value (..), (.=), (.:))
 import Data.Maybe (fromMaybe)
 import Data.Text (Text)
 import Prelude hiding (lookup)
 
+import qualified Data.Aeson as Aeson
+import qualified Data.Aeson.Types as Aeson
 import qualified Data.HashMap.Strict as HashMap
 
 type Path = [Text]
 
+-- A modification operation.
+data Modification
+  = Put Path Value
+  | Delete Path
+  deriving (Eq, Show)
+
+instance Aeson.ToJSON Modification where
+  toJSON (Put path value) = Aeson.object
+    [ "op" .= ("put" :: Text)
+    , "path" .= path
+    , "val" .= value
+    ]
+  toJSON (Delete path) = Aeson.object
+    [ "op" .= ("del" :: Text)
```

Comment:
I'd use the tag `"delete"` instead of `"del"`, to reduce the number of synonymous terms.

Reply:
I used `del` because it makes the journal entries up to the path line up in the journal file, making it much more parseable for the human eye. That may be an unnecessary concern though.

Reply:
It might not even, as the order in which the fields are serialized is unspecified.

Reply:
But it does. I get your point though, I'll change it.